File: 0090-subsets-ii/0090-subsets-ii.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n=len(nums)
        res=[]
        nums.sort()
        def subset(ind,ds,ans):
            ans.append(ds)            
            for i in xrange(ind,n):
                if i>ind and nums[i]==nums[i-1]:
                    continue
                #ds.append(nums[i])
                subset(i+1,ds+[nums[i]],ans)
                #ds.pop()

        subset(0,[],res)
        return res

File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        helper(0, new ArrayList<>(), result, nums);
        return  result;
    }
    
    private void helper(int index, List<Integer> path, List<List<Integer>> result, int[] nums){
        //Base Case
        result.add(new ArrayList<>(path));
        
        for (int i = index; i < nums.length; i++){
            if ((i > index) && (nums[i] == nums[i - 1])) continue;
            path.add(nums[i]);
            helper(i + 1, path, result, nums);
            path.remove(path.size()  - 1);
        }
    }
}